core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala

```diff
@@ -74,8 +74,13 @@ private[spark] abstract class BasePythonRunner[IN, OUT](
   private val reuseWorker = conf.getBoolean("spark.python.worker.reuse", true)
   // each python worker gets an equal part of the allocation. the worker pool will grow to the
   // number of concurrent tasks, which is determined by the number of cores in this executor.
-  private val memoryMb = conf.get(PYSPARK_EXECUTOR_MEMORY)
+  private val memoryMb = if (Utils.isWindows) {
```

Comment:
I see. I think the point of view is a bit different. What I was trying to do is that:
we declare this configuration is not supported on Windows, meaning we disable this configuration on Windows from the start, JVM side - because it's JVM to launch Python workers. So, I was trying to leave the control to JVM.

> It seems brittle to disable this on the JVM side and rely on it here. Can we also set a flag in the ImportError case and also check that here?

However, in a way, It's a bit odd to say it's brittle because we're already relying on that.

Reply:
I mean that it is brittle to try to use `resource` if the JVM has set the property. You handle the `ImportError`, but the JVM could set the request and Python would break again.

I think that this should not be entirely disabled on Windows. Resource requests to YARN or other schedulers should include this memory. The only feature that should be disabled is the resource limiting on the python side.

Reply:
> JVM could set the request

This is handled in JVM so it wouldn't break. `worker` itself is strongly coupled to JVM.

You mean that case when the client is in Windows machine and it uses a Unix-based clusters, right? I think this is what the fix already does - the `PythonRunner`s already are created at executor side and it wouldn't affect when the client is on Windows.

Reply:
> My point is that if resource can't be loaded for any reason, the code shouldn't fail. As it is, if resource can't be loaded then that is handled, but if the memory limit is set then the worker will still try to use it. That's what I think is brittle. There should be a flag for whether to attempt to use the resource API, based on whether it was loaded.

Ah, so the point is that the condition for the existence `resource` might not be clear - so we should have the flag to make it not failed in case. Yup, makes sense. Let me add a flag.

> If the worker operates as I described, then why make any changes on the JVM side?

I am making some changes on the JVM side so that we can explicitly disable on a certain condition For instance, if we don't make a change in JVM side, and only make the changes in Python `worker`. Later, we can have some other changes in JVM side referring this configuration - which can be problematic.

If we keep the configuration somehow in JVM side, it basically means we could have another status for this configuration, rather then just being disabled or enabled which we should take care of.

Reply:
@HyukjinKwon, what should the JVM side do differently if `resource` is not available?

I don't think it should do anything different. It should still allocate the python memory region when requesting resources from schedulers. The only difference is that python isn't self-limiting. Do you have an example of something that the JVM should change when running on Windows?

Reply:
For the current status, there's no diff. I'm just trying to match to what we have been usually doing - configuration control is usually made in JVM side and I propose to disallow this configuration on Windows. I think it's pretty minor point now because I added the flag as well ...

Reply:
There is no configuration to change needed on the JVM side.

The JVM should communicate to Python how much memory it is allocated. If Python can limit itself to that amount, then that's fine. If the JVM doesn't expect Python to be able to limit, why would it not tell Python how much memory it was allocated?

There is no benefit to making this change that I can see.

Reply:
I am doing this because:

1. The code consistency - usually the configuration is dealt with in JVM side if possible - otherwise we should send the configuration to worker side and the process it in Python worker side ideally. What we need to do is disable the configuration on a certain condition explicitly.

2. If we do it only in Python side, it's going to introduce the third status in JVM (`memoryMb`). When the configuration value exists in JVM (which means it's enabled) but the functionaility is disabled in Python side. Dealing with it will reduce the complexity.

Why are you so against about disabling it in JVM side? I don't see big downside of doing this. I added a flag as you said as well. For clarification, it's a rather minor point.

Reply:
> functionality is disabled in Python side

The only functionality that is disabled is limiting the memory space. The allocation for Python is still requested from resource managers. Setting the environment property tells python how much memory it was allocated, no matter how that is used or enforced.

> code consistency - usually the configuration is dealt with in JVM side if possible

The JVM is handling the setting by requesting that memory for python and passing on the amount requested to python. The fact that the python process can't limit doesn't affect how the JVM side should behave. This needlessly couples JVM and python behavior with an assumption that may not be true in the future.

> Why are you so against about disabling it in JVM side?

There is no benefit to disabling this. It is more code with no purpose and it makes assumptions about what python can or cannot do that aren't obvious.

What if pandas implements some method to spill to disk to limit memory consumption? Will implementers of that future feature know that the environment variable is not set when running in windows? This adds complexity for no benefit because it doesn't change either the resource allocation in the JVM or the behavior of the python process. It only avoids sending valuable information. I see no reason for it.

Reply:
BTW I agree with Ryan that this seems unnecessary. Also because the checks on either side are not the same (checks for Windows on the Java side, checks for a specific module on the Python side).

Reply:
Strictly we should move it into JVM rather then adding more controls at workers. Otherwise we will end up sending a bunch of data and environment variables into python workers. It already send the environment variable that indicates that this configuration is set and I was thinking what we should do is to disable it on certain condition rather then checking the package and skip in Python worker side.

I would rather remove python side's check then.

Reply:
If there are more conditions found to skip this, then we should do that later in JVM side if possible. If not, we can discuss there later. If it's absolutely required to skip in worker side, we can do it later and remove JVM side check.

Reply:
Documentation should be "For platforms where the `resource` API is available, python will limit its resource usage". The allocation on the JVM side is still the correct behavior.

Reply:
> The allocation on the JVM side is still the correct behavior.

Also, is it ever tested on Windows?

Reply:
> Strictly we should move it into JVM rather then adding more controls at workers.

There's a reason why the value is sent to the worker. It enables a feature that, when available, gives you better error information.

With the python code, if the app runs over the specified limit, you'll get an error from python saying it's using more memory than it should.

Without it, you'll get a generic error from the resource manager that your app exceeded its memory allocation, and you wont' know exactly what caused it (java? python? something else?).

Reply:
Thanks, Vanzin. However, it really brings complexity on the other hand. Maintaining codes for Windows actually are quite costly. We can just simply make it atomic if it works or not by disabling it on Windows rather then making the feature complicated by introducing another state.

Reply:
To be honest, I'm skeptical if it really works or not on Windows.

Reply:
I don't understand what you're getting at here.

The point Ryan and I are making is that there's no point in having this check in the Java code, because the Python code already handles it. It doesn't work on Windows, and the updated Python code handles that.

The only shortcoming here is passing an env variable to Python that won't be used if you're on Windows. That's really super trivial and not a big deal at all.

Reply:
> The point Ryan and I are making is that there's no point in having this check in the Java code, because the Python code already handles it. It doesn't work on Windows, and the updated Python code handles that.

My point is we should remove the Python condition. 

Also, I mean I wonder if this whole feature itself introduced at https://github.com/apache/spark/pull/21977 works or not since it looks not ever tested.

Reply:
@vanzin, I mean the first change I did at https://github.com/apache/spark/pull/23055/commits/2d3315a7dab429abc4d9ef5ed7f8f5484e8421f1

Reply:
I mean .. what I am disabling here is setting Python memory limit on Windows; looks Yarn side still can allocate more but not tested.

I'm thinking we should disable whole feature on Windows ideally but I couldn't either test it on Windows because I don't have Windows Yarn cluster (I only have one Windows machine that has dev environment).

I was trying to at least document that it doesn't work because it's going to work differently comparing to other OSes that don't have `resource` module (For current status, PySpark apps that uses Python worker do not work at all and we don't necessarily document it works on Windows). I think it's more correct to say it does not work because it's not tested (or at least just simply say it's not guaranteed on Windows).

For the reason that I prefer to check it on JVM side instead is, 
1. It's really usual to check it on JVM side when it's possible and make the worker simple. It could make it coupled to JVM but it's already strongly coupled

2. If Yarn code can also be tested, and if it doesn't work, it should also be disabled in JVM side.
If we can find a workaround on Windows, we can fix the Python worker and enable it.

3. If it's checked JVM side ahead, it reduces one state in JVM (`memoryMb`) is enabled in JVM side but Python skips it.

Reply:
I mean, even if it succeeds to allocate in Yarn and Python worker doesn't have the control on that, what's the point?

Reply:
BTW, i don't get why we should argue like this for this one. Isn't it a minor point?

Reply:
It's a minor point, but you seem to be the one making a big fuss about it.

Why is this code needed? It's not doing anything useful if you keep the python check around. The JVM doesn't understand exactly what Python supports of not, it's better to let the python code decide that.

You say we should disable the feature on Windows. The python-side changes already do that. We should not remove the extra memory requested from the resource manager just because you're running on Windows - you'll still need that memory, you'll just get a different error message if you end up using more than you requested.

> If we can find a workaround on Windows, we can fix the Python worker and enable it.

Exactly, and that should not require making any changes on the JVM side.

Reply:
I don't think I'm only the one tho. 

> Why is this code needed? 

I explain above multiple times. See above.

> it's not doing anything useful if you keep the python check around

See above. I want to delete them but added per review comment.

>  The JVM doesn't understand exactly what Python supports of not, it's better to let the python code decide that.

Not really. resource module is a Python builtin module that exists unix based system. It just does not exist in Windows.

> You say we should disable the feature on Windows. The python-side changes already do that. 

I explained above. See the first change I proposed 2d3315a. It relays on the environment variable.

> We should not remove the extra memory requested from the resource manager just because you're running on Windows - you'll still need that memory, you'll just get a different error message if you end up using more than you requested.

Yea, I know it would probably work. My question that is it ever tested? One failure case was found and it looks a bit odd that we document it works. It's not even tested and shall we make it simple rather then it make it work differently until it's tested?